### PR TITLE
dotCMS/core#23225 Allow user to edit page or master contentlet

### LINKED ISSUE #23225: Allow user to edit page or master contentlet
### Is your feature request related to a problem? Please describe.
As a dotCMS user, I want to be able to decide if I would like to edit a piece of content just for the page I'm editing or for the whole system.

### Describe the solution you'd like
When we know the contentlet is being used in more than one page, let's show the user a dialog, so it can choose if they want to edit the contentlet for the current page or for all the pages (master).

### Additional context
If the user selects "just for this page", we need to hit the endpoint to copy in place the contentlet and they open the editor for this one.

- [To know if the contentlet is being use in more than one page](https://github.com/dotcms/core/issues/23223)
- [To create a copy of the contentlet](https://github.com/dotcms/core/issues/23224)

### Mock
[Figma Prototype](https://www.figma.com/proto/trbwjLuf5F6KaU1rFucNeg/Portlets-%2F-Edit-Page?page-id=0%3A1&node-id=569%3A4774&viewport=4311%2C-717%2C0.42&scaling=min-zoom&starting-point-node-id=569%3A4774)

image

### Copy in place endpoint

```bash
curl --location --request PUT 'http://localhost:8080/api/v1/page/copyContent' \
--header 'Content-Type: application/json' \
--data-raw '{
    "containerId": "",
    "contentId": "",
    "pageId": "",
    "personalization": "",
    "relationType": "",
    "treeOrder": "",
    "variantId": ""
}'
```

### Acceptance criteria
- Show the dialog only when the content is used in multiples pages
- Allow the user to edit the contentlet for this page or all pages.

Updated by @fmontes

### PR DESCRIPTION
**Pull request by rjvelazco**
### Proposed Changes
* Allow user to edit page or master `contentlet`.

### Checklist
- [x] Tests
- [x] Translations
- [x] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots

https://user-images.githubusercontent.com/72418962/224394056-7191b69c-9170-436c-a795-0496141a0b7c.mov

### CODE REVIEW

File: core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        const requestOptions: DotRequestOptionsArgs = {
+            method: 'PUT',
+            body: {
+                ...body,
+                personalization: body.personalization || DEFAULT_PERSONALIZATION
+            },
+            url: `api/v1/page/copyContent`
+        };
+
+        return this.coreWebService.requestView(requestOptions).pipe(pluck('entity'), shareReplay());
```
Comment (oidacra) [inline core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts:35]: We will still be using coreWebServices to make a request? Not deprecated?

**Review by jcastro-dotcms — APPROVED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts [hunk 2]
```diff
@@ -715,7 +736,9 @@ export class DotEditContentHtmlService {
                     // because: https://github.com/dotCMS/core/issues/21818
 
                     setTimeout(() => {
-                        this.renderEditedContentlet(this.currentContentlet);
+                        this.currentMenuAction === DotContentletMenuAction.edit
+                            ? this.renderEditedContentlet(this.currentContentlet, contentlet)
+                            : this.renderEditedContentlet(this.currentContentlet);
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts:741]: Are you sure we need this ternary?

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/dot-edit-content.component.ts [hunk 3]
```diff
@@ -682,4 +721,54 @@ export class DotEditContentComponent implements OnInit, OnDestroy {
...
+        inode
+    }: {
+        copyContent: DotCopyContent;
+        inode: string;
+    }): void {
+        const ref = this.dialogService.open(DotBinaryOptionSelectorComponent, {
+            header: this.dotMessageService.get('Edit-Content'),
+            width: '37rem',
+            data: { options: this.CONTENT_EDIT_OPTIONS },
+            style: { 'border-radius': '16px' },
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/dot-edit-content.component.ts:741]: Remove this, will be doing this globally for all the dialogs.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-templates/dot-template-create-edit/dot-template-new/dot-template-new.component.ts [hunk 4]
```diff
@@ -5,28 +5,49 @@ import { DialogService } from 'primeng/dynamicdialog';
...
     ) {}
 
     ngOnInit(): void {
-        const ref = this.dialogService.open(DotTemplateSelectorComponent, {
+        const ref = this.dialogService.open(DotBinaryOptionSelectorComponent, {
             header: this.dotMessageService.get('templates.select.template.title'),
-            width: '37rem'
+            width: '37rem',
+            data: { options: this.options },
+            style: { 'border-radius': '16px' },
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-templates/dot-template-create-edit/dot-template-new/dot-template-new.component.ts:46]: +1

File: core-web/apps/dotcms-ui/src/app/portlets/shared/dot-binary-option-selector/dot-binary-option-selector.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
+import { DebugElement } from '@angular/core';
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { By } from '@angular/platform-browser';
+// import { By } from '@angular/platform-browser';
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/shared/dot-binary-option-selector/dot-binary-option-selector.component.spec.ts:4]: clean.

File: core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+export const DEFAULT_PERSONALIZATION = 'dot:default';
+
+@Injectable({
+    providedIn: 'root'
+})
+export class DotCopyContentService {
+    constructor(private readonly coreWebService: CoreWebService) {}
+
+    /**
+     *
```
Comment (fmontes) [inline core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts:19]: Add desc.

File: core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts (hunk 1 above)
Comment (fmontes) [reply core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts:35]: Yes... deprecated... update to http module @oidacra can help you with this.

**Review by fmontes — APPROVED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/dot-edit-content.component.ts (hunk 3 above)
Comment (rjvelazco) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/dot-edit-content.component.ts:741]: Great, I'll do it!

File: core-web/apps/dotcms-ui/src/app/portlets/dot-templates/dot-template-create-edit/dot-template-new/dot-template-new.component.ts (hunk 4 above)
Comment (rjvelazco) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-templates/dot-template-create-edit/dot-template-new/dot-template-new.component.ts:46]: Great, I'll do it!

File: core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts (hunk 6 above)
Comment (rjvelazco) [reply core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts:19]: I'll do it!

File: core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts (hunk 1 above)
Comment (rjvelazco) [reply core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts:35]: I'll mark it as `@deprecated` and  use the `httpClientModule`.

**Review by oidacra — APPROVED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts [hunk 7]
```diff
@@ -178,6 +202,8 @@ export class DotEditContentHtmlService {
 
         contenletEl.remove();
 
+        this.savePage(this.getContentModel()).subscribe();
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts:205]: take1?

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts [hunk 8]
```diff
@@ -178,6 +202,8 @@ export class DotEditContentHtmlService {
...
+
         this.pageModel$.next({
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/dot-edit-content-html/dot-edit-content-html.service.ts:207]: I wonder if we should wait the save before doing this.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/dot-edit-content.component.ts [hunk 9]
```diff
@@ -682,4 +664,62 @@ export class DotEditContentComponent implements OnInit, OnDestroy {
...
+            }
+
+            this.CONTENT_EDIT_OPTIONS.option1.value === value
+                ? this.copyContentAndEdit(copyContent)
+                : this.editContentlet(inode);
+        });
+    }
+
+    /**
+     *
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/dot-edit-content.component.ts:700]: Doc... copilot it!

File: core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts (hunk 6 above)
Comment (fmontes) [reply core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts:21]: Remove the extra line, not sure if the parse will parse like thie.

File: core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+     * @return {*}  {Observable<DotCMSContentlet>}
+     * @memberof DotCopyContentService
+     */
+    copyContentInPage(data: DotCopyContent): Observable<DotCMSContentlet> {
+        const body = {
+            ...data,
+            personalization: data?.personalization || DEFAULT_PERSONALIZATION
+        };
+
+        return this.http.put(API_ENDPOINT, body).pipe(shareReplay(), pluck('entity'));
```
Comment (fmontes) [inline core-web/libs/data-access/src/lib/dot-copy-content/dot-copy-content.service.ts:33]: What is the `shareReplay` for?

File: core-web/libs/dotcms-js/src/lib/core/core-web.service.ts [hunk 11]
```diff
@@ -50,6 +50,12 @@ export interface DotRequestOptionsArgs {
     };
 }
 
+/**
+ * @deprecated
+ *
+ * @export
+ * @class CoreWebService
+ */
```
Comment (fmontes) [inline core-web/libs/dotcms-js/src/lib/core/core-web.service.ts:58]:
```
/**
 * Request data from dotCMS endpoints
 * @deprecated use Angular HttpClient instead
 * @export
 * @class CoreWebService
 */
 ```

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/dotcms-js/src/lib/core/core-web.service.ts (hunk 11 above)
Comment (rjvelazco) [reply core-web/libs/dotcms-js/src/lib/core/core-web.service.ts:58]: I'll add it!

**Pull request merged**